### ISSUE TITLE
errno shouldn't propagate into user's code

### ISSUE THREAD

**Issue by light2yellow** [Snaipe/Criterion#246]
Criterion either doesn't check for `errno` and doesn't fail somewhere, or it doesn't reset it before running user's testcase. Either way, as a user I want to assume my `errno` is not being set by the framework, if only by my fixtures, i.e. own code.

Repro:
```c
% cat example.c   
                 
#include <stdio.h>
#include <errno.h>
#include <criterion/criterion.h>
#include <criterion/logging.h>
Test(suite_name, test_name) {
        cr_log_info("errno: %s(%d)", strerror(errno), errno);
}
```
```
% gcc example.c -lcriterion
% ./a.out --verbose
[----] Criterion v2.3.2
[====] Running 1 test from suite_name:
[RUN ] suite_name::test_name
[----] errno: Resource temporarily unavailable(11)
```

**Comment by light2yellow** [comment]
Note that compiling with `--std=c11` yields the same result. I'm saying this because `errno` is thread-local since C11, and Criterion might run in multiple threads (I don't know if it's true). So, it's not leaking from some other thread.

**Comment by Snaipe** [comment]
`errno` doesn't have to have any sort of required value when a function succeeds. In particular, consider that a function that succeeds might call functions that set errno during normal execution (this is true in particular for the error code you mentioned, `EAGAIN`, which literally means "I'm not ready yet, try to call that function again later").

Criterion spawns a completely new process, so the errno comes from the initialization process (either from Criterion or the pre-main libc code). I could explicitely set errno to 0 before running the test function, but checking that errno is nonzero after the test would be asking for inconsistent tests and wrong behaviour.

In short: if you need to test errno, do so in an assertion.

**Comment by light2yellow** [comment]
> In short: if you need to test errno, do so in an assertion.

Yes, that what I'm doing. But do you think leaving errno set to something is a good practice? In my opinion, it kind of diminishes all the effort taken to isolate a testcase, leaving such a simple sink. It's up to you, of course. It takes just one line in my own code for each case, so, nothing that makes this library unusable. Feel free to close if you consider it a "won't fix".

**Comment by Snaipe** [comment]
Well, it's not really about what I think is good practice, but more about what the spec says. There are countless APIs, [starting with the libc itself](https://www.gnu.org/software/libc/manual/html_node/Checking-for-Errors.html), that set errno to a nonzero value even when the function succeeds (see EAGAIN, EWOULDBLOCK, EINTR, ...).

**Comment by light2yellow** [comment]
> Warning: Many library functions may set errno to some meaningless non-zero value even if they did not encounter any errors, and even if they return error codes directly. Therefore, it is usually incorrect to check whether an error occurred by inspecting the value of errno. The proper way to check for error is documented for each function. 

Ah, okay, thanks. So, IIUC the general practise is to check for the return value, and then for the errno if something.
